proposals/0034-collection-references.md

```diff
@@ -0,0 +1,214 @@
+  const posts = await getCollection('blog');
+  return posts.map(p => ({
+    params: { slug: p.slug },
+    props: post,
+  }))
+}
+
+const blogPost = Astro.props;
+const author = await getEntry(blogPost.data.author);
+const { Content } = await blogPost.render();
+---
+
+<h1>{blogPost.data.title}</h1>
+<p>Author: {author.data.name}</p>
+<Content />
+```
+
+# Background & Motivation
+
+Content collections were designed to be configured and queried individually. But as users have started adapting collections to more complex patterns, there are compelling use cases to "reference" one collection entry from another:
+
+- Reference "related posts" within a collection of blog posts.
+- Create a collection of authors, and reference those authors from a collection of documentation pages.
+- Create a collection of images with reusable alt text, and reference those images for article banners.
+- Create a collection of tags with display text or icons, and reference those tags from a collection of blog posts.
+
+These use cases span data collection -> content collection references, content -> content references, and even references within the same collection. These make a "reference" primitive compelling as collections types grow.
+
+# Goals
+
+- Introduce an API to reference collection entries from another collection, regardless of the collection type (content or data).
+- Support references to other entries within the same collection.
+- Consider Both one-to-one and one-to-many relationships between content and data (ex. allow passing a list of author IDs in your frontmatter).
+
+
+# Non-Goals
+
+- **First-class helpers for many-to-many references.** In other words, if a blog post has authors, Astro will not help retrieve blog posts by author. This will require manual querying for all blog posts, and filtering to find blog posts containing a particular author.
+
+# Detailed Design
+
+The `reference()` utility receives the collection name as a string, and validates this string at build-time using a Zod transform. This transform does _not_ attempt to import the referenced object directly, instead returning the referenced identifier and collection name after validating.
+
+## Configuration
+
+This example configures a `blog` collection with a few properties that use references:
+- `banner` - entry in the `banners` data collection, containing image asset srcs and reusable alt text
+- `relatedPosts` - list of entries in the current `blog` collection
+- `authors` - list of entries in the `authors` data collection, containing author information
+
+```ts
+import { defineCollection, reference, z } from 'astro:content';
+
+const blog = defineCollection({
+  schema: z.object({
+    title: z.string(),
+    banner: reference('banners'),
+    authors: z.array(reference('authors')),
+    relatedPosts: z.array(reference('blog')),
+  })
+});
+
+const banners = defineCollection({
+  type: 'data',
+  schema: ({ image }) => z.object({
```

Comment:
Forgive me if this was already discussed, but seeing this all laid out now: it feels like a leaky implementation detail that `image` needs to come from a special schema function, but `z` and `reference` don't. There's no reason for this other than "the implementation of image() is special and needs this", correct?

Thoughts on passing `z` and `reference` to the schema function as well, and beginning to push users to always the function signature API instead? Looking at that API in the snippet below and it does seem more consistent to pass all schema helpers the same way. It's also a fairly small change that shouldn't be too disruptive to developers, and we could do it in a backwards-compatible way.


```diff
- import { defineCollection, reference, z } from 'astro:content';
+ import { defineCollection } from 'astro:content';

const blog = defineCollection({
-  schema: z.object({
+  schema: ({z, reference}) => z.object({
    title: z.string(),
    banner: reference('banners'),
    authors: z.array(reference('authors')),
    relatedPosts: z.array(reference('blog')),
  }),
});
const banners = defineCollection({
  type: 'data',
-  schema: z.object({
+  schema: ({z, image}) => z.object({
    src: image(),
    alt: z.string(),
  })
});
const authors = defineCollection({
  type: 'data',
-  schema: z.object({
+  schema: ({z}) => z.object({
    name: z.string(),
  })
});
export const collections = { blog, banners, authors };
```

Not a hill I'll die on, but I think it's worth the change IMO in the name of consistency and removing leaky implementation details!

Reply:
I think that image is special (it's contextual) *is* a good reason for it to be accessed from a different place. It's not really an implementation detail, we couldn't have implemented it differently. 

Aside from that, given the discussions about "Content Collections v2" it seems like unnecessary churn to move `z` now when defining schemas might look quite different in the near future anyways. `reference` I feel less strongly about. Having it go through the function args and the function args representing "special Astro schema helpers" seems fine to me.

Reply:
I agree with Matthew on moving `z` here. I think allowing `reference()` to come from the arrow function _or_ an import is an okay compromise to meet users where they are

Reply:
Discussed with docs and core offline. To avoid thrashing in our API, we decided _just_ making `reference()` a top-level import for launch makes the most sense. We can explore other import options by user demand.